src/components/input/input.js

```diff
@@ -650,7 +651,23 @@ function mdMaxlengthDirective($animate, $mdUtil) {
       // Using the $validators for triggering the update works very well.
       renderCharCount();
 
-      return ( modelValue || element.val() || viewValue || '' ).length <= maxlength;
+      var elementVal = element.val() || viewValue;
+      if (elementVal === undefined || elementVal === null) {
+        elementVal = '';
+      }
+      elementVal = ngTrim && !isPasswordInput && angular.isString(elementVal) ? elementVal.trim() : elementVal;
+      // Force the value into a string since it may be a number,
+      // which does not have a length property.
+      return String(elementVal).length <= maxlength;
+    };
+
+    ngModelCtrl.$isEmpty = function(value) {
+      var elementVal = value;
+      if (elementVal === undefined || elementVal === null) {
+        elementVal = '';
+      }
+      elementVal = ngTrim && !isPasswordInput && angular.isString(elementVal) ? elementVal.trim() : elementVal;
+      return String(elementVal).length === 0;
```

Comment:
Nit: You can make a re-usable function for calculating the length of a value (which is used in the validator, `$isEmpty()` and `renderCharCount()`):

```js
function caclulateLength(value) {
  if (value == null) {
    value = '';
  } else {
    value = ngTrim && !isPasswordInput && angular.isString(value) ? value.trim() : value;
  }
  return String(value).length;
}
```

Reply:
I had a feeling that this was needed. Thanks for confirming it. Much cleaner 😃